subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/IncrementalCompileSourceProcessorCache.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.language.nativeplatform.internal.incremental;
+
+import com.google.common.base.Objects;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
+
+public class IncrementalCompileSourceProcessorCache {
+    private final Map<File, IncrementalCompileFilesFactory.FileDetails> fileBasedCache = new HashMap<File, IncrementalCompileFilesFactory.FileDetails>();
+    private final Map<Integer, IncrementalCompileFilesFactory.FileDetails> macroBasedCache = new HashMap<Integer, IncrementalCompileFilesFactory.FileDetails>();
+    private final Object lock = new Object();
+
+    public IncrementalCompileFilesFactory.FileDetails get(File file, Integer macroHash) {
```

Comment:
I don't think this is right. The include graph for a given file also takes as input the include path.

For example, let's say I have:

- `include1/a.h`
- `include2/a.h`
- `common/b.h` contains `#include <a.h>`

Then I have 2 projects:

Project1 has `src/main/cpp/app1.cpp` contains `#include <b.h>` and compiled with include path `common:include1`. This means `app1.cpp` depends on `common/b.h` and `include1/a.h`

Project2 has `src/main/cpp/app2.cpp` also contains `#include <b.h>` and compiled with include path `common:include2`. This means `app2.cpp` depends on `common/b.h` and `include2/a.h`.

However, with the implementation in this PR, we're going to cache the result of analysing `common/b.h` for the first source file that sees it, and then reuse the result for the other, which will give the wrong dependencies.

There's a similar thing going on with the macro functions.

Reply:
I think a refinement to using the full set of visible macros (the macro hash) would be to use the intersection of the visible macros and those macros actually used to resolve include files. Otherwise, we'll get misses if one source file includes an include file that defines some macros (and what include file doesn't?) and then includes some other include file that we've already analysed with a different set of visible macros, whereas we'd get a hit if we ignored those macros that are irrelevant to calculating the include file graph.

This seems like a very common pattern in code that isn't artificial code in a test build, and we may be seeing an improvement due to this caching in our test build because it is using artificial code.

Reply:
Thanks for this comments, I had a feeling I was oversimplifying the problem. Let me see what I can do to consider include roots as well.